### PR TITLE
Add rewrite for argmax/argmin of monotonic functions

### LINKED ISSUE #1851: Rewrite `argmax(f(x)` -> argmax(x)` for monotonic f
### Description

Similarly `max(f(x)) -> f(max(x))`

We can also convert (arg)max to (arg)min for monotonic decreasing functions

### PR DESCRIPTION
**Pull request by Jasjeet-Singh-S**
# Add rewrite for argmax/argmin/max/min of monotonic functions

Closes #1851

## Summary

This PR implements graph rewrites that optimize `argmax`/`argmin`/`max`/`min` operations applied to monotonic functions by eliminating unnecessary function evaluations or moving them outside reductions.

## Motivation

Computing `argmax(exp(x))` or `max(exp(x))` is wasteful because the exponential computation doesn't affect which index has the maximum value or the relative ordering - we only care about the ordering relationship. Since monotonic functions preserve ordering, we can skip expensive function applications entirely for argmax/argmin, or move them outside the reduction for max/min.

### Monotonicity Attribute System

Added a `Monotonicity` enum to `pytensor.scalar.basic` to mark scalar operations:
- `Monotonicity.INCREASING` (+1): Function preserves ordering
- `Monotonicity.DECREASING` (-1): Function reverses ordering  
- `Monotonicity.NONMONOTONIC` (0): Default for unmarked operations

This property is reusable across PyTensor for future optimizations.

### New Rewrites

**1. `local_argmax_argmin_monotonic`** - for argmax/argmin operations

Transformation paths based on function monotonicity:

**Monotonically Increasing Functions:**
- `argmax(f(x)) → argmax(x)` 
- `argmin(f(x)) → argmin(x)`

**Monotonically Decreasing Functions:**
- `argmax(f(x)) → argmin(x)`
- `argmin(f(x)) → argmax(x)`

**2. `local_max_min_monotonic`** - for max/min operations

Moves monotonic functions outside reductions:

**Monotonically Increasing Functions:**
- `max(f(x)) → f(max(x))`
- `min(f(x)) → f(min(x))`

**Monotonically Decreasing Functions:**
- `max(f(x)) → f(min(x))`
- `min(f(x)) → f(max(x))`

### Supported Functions

**Increasing (18 ops):** `Exp`, `Exp2`, `Expm1`, `Log`, `Log2`, `Log10`, `Log1p`, `Sqrt`, `Deg2Rad`, `Rad2Deg`, `ArcSin`, `Tan`, `ArcTan`, `ArcCosh`, `Sinh`, `ArcSinh`, `Tanh`, `ArcTanh`

**Decreasing (3 ops):** `Neg`, `Reciprocal`, `ArcCos`

## Key Features

- **Extensible design**: Monotonicity is a reusable scalar Op property
- **Handles PyTensor's internal representation**: Correctly processes `argmin` (internally `Argmax(Neg(...))`)
- **Preserves axis parameter**: Works with different axis specifications (`None`, `0`, `-1`, etc.)
- **Robust pattern matching**: Uses `Elemwise` wrapper detection to identify scalar operations
- **Stack trace preservation**: Maintains debugging information via `copy_stack_trace`
- **Comprehensive coverage**: Handles both value-returning (max/min) and index-returning (argmax/argmin) reductions

### `pytensor/scalar/basic.py`
- Added `Monotonicity` enum (`IntEnum`) with values `INCREASING`, `NONMONOTONIC`, `DECREASING`
- Added `monotonicity` class attribute to `ScalarOp` (defaults to `NONMONOTONIC`)
- Marked 21 scalar operations with appropriate monotonicity values

### `pytensor/tensor/rewriting/math.py`
- Implemented `_is_argmin()` helper to detect argmin patterns (`Argmax(Neg(...))` representation)
- Implemented `local_argmax_argmin_monotonic()` rewriter with `@register_canonicalize`
- Implemented `local_max_min_monotonic()` rewriter with `@register_canonicalize`
- Rewrites check `scalar_op.monotonicity` attribute instead of hardcoded tuples

### `tests/tensor/rewriting/test_math.py`
- Added `TestArgmaxArgminMonotonic` class with 8 test methods:
  - `test_argmax_increasing_functions` - Argmax with increasing functions
  - `test_argmin_increasing_functions` - Argmin with increasing functions
  - `test_argmax_decreasing_functions` - Argmax with decreasing functions (flips to argmin)
  - `test_argmin_decreasing_functions` - Argmin with decreasing functions (flips to argmax)
  - `test_max_increasing_functions` - Max with increasing functions
  - `test_min_increasing_functions` - Min with increasing functions
  - `test_max_decreasing_functions` - Max with decreasing functions
  - `test_min_decreasing_functions` - Min with decreasing functions
- All tests parametrized over `axis ∈ [None, 0, -1]`
- Tests verify both numerical correctness and graph structure optimization
- Uses `arccos` for decreasing tests to avoid confusion with argmin's internal `Neg` representation

### Argmax/Argmin
```python
import pytensor.tensor as pt

x = pt.vector('x')
y_argmax = pt.argmax(pt.exp(x))  # Before: exp(x) then argmax
                                  # After: argmax(x) directly
y_argmin = pt.argmin(pt.exp(x))  # Before: exp(x) then argmin
                                  # After: argmin(x) directly
```

### Max/Min
```python
import pytensor.tensor as pt

x = pt.vector('x')
y_max = pt.max(pt.exp(x))  # Before: exp(x) then max
                            # After: exp(max(x))
y_min = pt.min(pt.exp(x))  # Before: exp(x) then min
                            # After: exp(min(x))

# Reduces N exponential ops to 1
```

## Performance Impact

Significant speedups when:
- Computing argmax/argmin/max/min of expensive monotonic transformations (exp, log, sqrt, etc.)
- Working with large arrays where:
  - **argmax/argmin**: Eliminates all function evaluations
  - **max/min**: Reduces N expensive ops to 1 (N = reduction dimension size)
- Monotonic function is the computational bottleneck

## Testing

**All 24 tests pass:**
- 12 argmax/argmin tests (4 methods × 3 axes)
- 12 max/min tests (4 methods × 3 axes)
- 21 monotonic functions tested (18 increasing, 3 decreasing)
- Multiple axis configurations (`None`, `0`, `-1`)
- Numerical correctness verified
- Graph structure validated

**Edge cases handled:**
- PyTensor's `Argmax(Neg(...))` representation for `argmin`
- Broadcasting and dimension handling
- Proper flipping between argmax/argmin for decreasing functions
- Moving functions outside reductions for max/min

### CODE REVIEW

File: pytensor/tensor/rewriting/math.py [hunk 1]
```diff
@@ -121,6 +125,14 @@
     TensorVariable,
 )
```
Comment (ricardoV94) [inline pytensor/tensor/rewriting/math.py:127]: Let's add these properties to the Scalar Op directly. They are going to be useful in other contexts

Comment (Jasjeet-Singh-S) [reply pytensor/tensor/rewriting/math.py:127]: Should i implement this by using a class attribute (-1,0,1 for decreasing, non-monotonic, increasing) or should i make a decorator?

Comment (ricardoV94) [reply pytensor/tensor/rewriting/math.py:127]: Let's do a class attribute. We can use an Enum perhaps

**Comment by Jasjeet-Singh-S** [comment]
Hi @ricardoV94 , i was curious as to what is the procedure for moving forward with these PR's since its my first time with open source. Are there any more issues i need to address?

**Comment by ricardoV94** [comment]
@Jasjeet-Singh-S I just didn't have the time to check yet, apologies.

**Comment by Jasjeet-Singh-S** [comment]
Hi @ricardoV94 , following up on this PR. Now that #2053  has been merged, the Monotonicity enum infrastructure I added in this PR is superseded by Jesse's monotonic_increasing/monotonic_decreasing properties on UnaryScalarOp. However, the actual rewrite logic — local_argmax_argmin_monotonic and local_max_min_monotonic — is not covered by #2053  and is still unimplemented in the codebase.
I'd like to rebase this PR and update the implementation to use the new property system from #2053  rather than my own enum. Would the rewrite logic still be welcome as a follow-up contribution? Happy to update as soon as I get confirmation the direction is right.
